core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2066,7 +2066,7 @@ class KafkaApis(val requestChannel: RequestChannel,
           .setErrorCode(Errors.NOT_CONTROLLER.code))
       }
       sendResponseCallback(results)
-    } else if (!config.deleteTopicEnable) {
+    } else if (!zkSupport.controller.topicDeletionManager.isDeleteTopicEnabled) {
```

Comment:
What happens to this line with KIP-500?

Reply:
@CCisGG 
The toggle is static in the upstream, i.e. doesn't change after deployment, as it only refers to the config.

Our LI modification basically adds a ZK listener to a specific ZNode, and changes the behavior when the ZNode is modified (although this line did not incorporate the change).

However, in KIP-500, we no longer have ZK, which means the path to notify "all the brokers" that "delete topic enable should be changed" is eliminated.

The deletion behavior would remain the same as the upstream after KIP-500 without further modification, but just we will lose the path to change it at the run time.

Reply:
Wouldn't changing this cause it doesn't read from static config anymore?

Reply:
@CCisGG you're right, and that's the whole point of the dynamic flag change feature.   If it uses the static config, there's no path to change it at run time. And in KIP-500, such implementation referencing the state would be equivalent to referencing the static config, as the TopicDeletionManager initiates the value from config and current there's no API to change it at run time.

Reply:
1. It seems the write and read of this variable could land on different threads, thus it needs to be a volatile variable. 
2. In the current code base, it seems this variable is only loaded during the initialization of the TopicDeletionManager, which means any further updates on ZK won't be picked up. When this code was first added, there was probably the ZK listener logic but it seems we've dropped that along the way. 
Maybe it's time to completely drop this zk flag.

Reply:
@gitlw 

1. Good catch! Fixed it to make it `@volatile`
2. The listener is still there
    * [ZkHandler that monitors the path adn sends out the controller event](https://github.com/linkedin/kafka/blob/3.0-li/core/src/main/scala/kafka/controller/KafkaController.scala#L3089)
    * [The controller event](https://github.com/linkedin/kafka/blob/3.0-li/core/src/main/scala/kafka/controller/KafkaController.scala#L3001)
    * [The body of the update](https://github.com/linkedin/kafka/blob/3.0-li/core/src/main/scala/kafka/controller/KafkaController.scala#L2183)